src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,122 @@
+    private static final double GROW_STEP = 1.5;
+    private Object[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = new Object[INITIAL_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        sizeCheck();
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+            return;
+        }
+        indexCheck(index);
+        sizeCheck();
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T get(int index) {
-        return null;
+        indexCheck(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        indexCheck(index);
+        elementData[index] = value;
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T remove(int index) {
-        return null;
+        indexCheck(index);
+        Object removedElement = elementData[index];
+        if (size - 1 - index >= 0) {
+            System.arraycopy(elementData, index + 1, elementData, index, size - 1 - index);
+        }
+        size--;
+        return (T) removedElement;
     }
 
     @Override
     public T remove(T element) {
+        int index = indexOf(element);
+        if (index >= 0) {
+            System.arraycopy(elementData, index + 1, elementData, index, size - index - 1);
+            size--;
+            elementData[size] = null;
+            return element;
+        }
         return null;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public void indexCheck(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index = " + index);
+        }
+    }
+
+    public void sizeCheck() {
+        if (size == elementData.length) {
+            ensureCapacity();
+        }
+    }
+
+    public void ensureCapacity() {
+        int newIncreasedCapacity = (int) (elementData.length * GROW_STEP);
+        elementData = Arrays.copyOf(elementData, newIncreasedCapacity);
+    }
+
+    public int indexOf(T element) {
+        for (int i = 0; i < size; i++) {
+            //don't know how to improve
+            if (element == null) {
+                if (element == elementData[i]) {
+                    return i;
+                }
+                continue;
+            }
+            if (element.equals(elementData[i])) {
+                return i;
+            }
```

Comment:
you can use one if element == elements[i] || element != null && element.equals(elements[i])